### ISSUE TITLE
EarlyStopping TensorFlow 2.0 discrepancy

### ISSUE THREAD

**Issue by arjun-majumdar** [tensorflow/tensorflow#36241]
I am running a code using Python 3.7.5 with TensorFlow 2.0 for MNIST classification.
I am using **EarlyStopping** from TensorFlow 2.0 and the callback I have for it is:

    callbacks = [
                 tf.keras.callbacks.EarlyStopping(
                     monitor='val_loss', patience = 3,
                     min_delta=0.001
                 )
    ]

According to [EarlyStopping - TensorFlow 2.0][1] page, the definition of **min_delta** parameter is as follows:

**min_delta: Minimum change in the monitored quantity to qualify as an
improvement, i.e. an absolute change of less than min_delta, will
count as no improvement.**

> Train on 60000 samples, validate on 10000 samples
> 
> Epoch 1/15 60000/60000 [==============================] - 10s
> 173us/sample - loss: 0.2040 - accuracy: 0.9391 - val_loss: 0.1117 -
> val_accuracy: 0.9648
> 
> Epoch 2/15 60000/60000 [==============================] - 9s
> 150us/sample - loss: 0.0845 - accuracy: 0.9736 - val_loss: 0.0801 -
> val_accuracy: 0.9748
> 
> Epoch 3/15 60000/60000 [==============================] - 9s
> 151us/sample - loss: 0.0574 - accuracy: 0.9817 - val_loss: 0.0709 -
> val_accuracy: 0.9795
> 
> Epoch 4/15 60000/60000 [==============================] - 9s
> 149us/sample - loss: 0.0434 - accuracy: 0.9858 - val_loss: 0.0787 -
> val_accuracy: 0.9761
> 
> Epoch 5/15 60000/60000 [==============================] - 9s
> 151us/sample - loss: 0.0331 - accuracy: 0.9893 - val_loss: 0.0644 -
> val_accuracy: 0.9808
> 
> Epoch 6/15 60000/60000 [==============================] - 9s
> 150us/sample - loss: 0.0275 - accuracy: 0.9910 - val_loss: 0.0873 -
> val_accuracy: 0.9779
> 
> Epoch 7/15 60000/60000 [==============================] - 9s
> 151us/sample - loss: 0.0232 - accuracy: 0.9921 - val_loss: 0.0746 -
> val_accuracy: 0.9805
> 
> Epoch 8/15 60000/60000 [==============================] - 9s
> 151us/sample - loss: 0.0188 - accuracy: 0.9936 - val_loss: 0.1088 -
> val_accuracy: 0.9748

Now if I look at the last three epochs viz., epochs 6, 7, and 8 and look at the validation loss ('val_loss'), their values are:

0.0688, 0.0843 and 0.0847.

And the differences between consecutive 3 terms are: 0.0155, 0.0004. But isn't the first difference greater than 'min_delta' as defined in the callback. 

The code I came up with for EarlyStopping is as follows:

    # numpy array to hold last 'patience = 3' values-
    pv = [0.0688, 0.0843, 0.0847]
    
    # numpy array to compute differences between consecutive elements in 'pv'-
    differences = np.diff(pv, n=1)
    
    differences
    # array([0.0155, 0.0004])
    
    
    # minimum change required for monitored metric's improvement-
    min_delta = 0.001
    
    # Check whether the consecutive differences is greater than 'min_delta' parameter-
    check = differences > min_delta
    
    check
    # array([ True,  False])
    
    # Condition to see whether all 3 'val_loss' differences are less than 'min_delta'
    # for training to stop since EarlyStopping is called-
    if np.all(check == False):
        print("Stop Training - EarlyStopping is called")
        # stop training 

But according to the 'val_loss', the differences between the not ALL of the 3 last epochs are greater than 'min_delta' of 0.001. For example, the first difference is greater than 0.001 (0.0843 - 0.0688) while the second difference is less than 0.001 (0.0847 - 0.0843).

Also, according to **patience** parameter definition of "EarlyStopping":

**patience: Number of epochs with no improvement after which training will be stopped.**

So, EarlyStopping should be called when there is no improvement for 'val_loss' for 3 consecutive epochs where the absolute change of less than 'min_delta' does not count as improvement.

Then why is EarlyStopping called?

Another example is:

Train on 60000 samples, validate on 10000 samples
Epoch 1/30
60000/60000 [==============================] - 11s 179us/sample - loss: 0.2032 - accuracy: 0.9399 - val_loss: 0.1195 - val_accuracy: 0.9626
Epoch 2/30
60000/60000 [==============================] - 9s 155us/sample - loss: 0.0852 - accuracy: 0.9732 - val_loss: 0.0891 - val_accuracy: 0.9712
Epoch 3/30
60000/60000 [==============================] - 9s 154us/sample - loss: 0.0552 - accuracy: 0.9830 - val_loss: 0.0897 - val_accuracy: 0.9723
Epoch 4/30
60000/60000 [==============================] - 9s 153us/sample - loss: 0.0460 - accuracy: 0.9850 - val_loss: 0.0746 - val_accuracy: 0.9794
Epoch 5/30
60000/60000 [==============================] - 9s 152us/sample - loss: 0.0325 - accuracy: 0.9894 - val_loss: 0.0787 - val_accuracy: 0.9774
Epoch 6/30
60000/60000 [==============================] - 9s 154us/sample - loss: 0.0267 - accuracy: 0.9913 - val_loss: 0.0736 - val_accuracy: 0.9782
Epoch 7/30
60000/60000 [==============================] - 9s 153us/sample - loss: 0.0250 - accuracy: 0.9920 - val_loss: 0.0857 - val_accuracy: 0.9777
Epoch 8/30
60000/60000 [==============================] - 9s 153us/sample - loss: 0.0210 - accuracy: 0.9929 - val_loss: 0.0808 - val_accuracy: 0.9790
Epoch 9/30
60000/60000 [==============================] - 9s 153us/sample - loss: 0.0179 - accuracy: 0.9937 - val_loss: 0.1043 - val_accuracy: 0.9768

Here, the last 3 'val_loss' are:
0.08566837142373625, 0.08082119292882962 and 0.10430033170154183.

And the consecutive differences between the 2 adjacent 'val_loss' are:
0.00484718 and 0.02347914

Now, both of these differences are greater than 'min_delta' of 0.001. But, 'EarlyStopping' again halts training.

What am I not getting?

  [1]: https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/EarlyStopping

The complete code GitHub URL:
https://github.com/arjun-majumdar/Lottery_Ticket_Hypothesis-TensorFlow_2/blob/master/tfmot_sparsity_experiment.ipynb

**Comment by amahendrakar** [comment]
@arjun-majumdar,
In order to expedite the trouble-shooting process, could you please provide the complete code to reproduce the issue reported here. Thanks!

**Comment by arjun-majumdar** [comment]
@amahendrakar , added GitHub repo URL in post

https://github.com/arjun-majumdar/Lottery_Ticket_Hypothesis-TensorFlow_2/blob/master/tfmot_sparsity_experiment.ipynb

**Comment by bfs15** [comment]
It is because the delta is calculated from the best value, on your second example it would be epoch 6 with val_loss 0.0736.
The next 3 epochs don't get better values than that. 0.0856, 0.0808 and 0.1043 are all greater than 0.0736 so they showed no improvement. The purpose is to improve on your best epoch not the latest.

**Comment by amahendrakar** [comment]
@arjun-majumdar,
Could you please verify if @bfs15's answer resolves your query?  Thanks!

**Comment by arjun-majumdar** [comment]
Thanks, it solves my problem
